### PR TITLE
Improve manual Size Mode handling in the editor

### LINKED ISSUE #7464: Respect `manual` size mode for sprites in the editor
**Describe the bug (REQUIRED)**
 `manual` size mode in the editor should be respected. Editor should set current size when user pick `Manual` and save it in sprite's proto file.

**To Reproduce (REQUIRED)**

1. Change sprite size to manual
2. for ex. sprite is 32x32
3. I see in width and height 32x32
4. I change image in atlas to 100x100 with the same name
5. return to the go with sprite sprite and now it’s 100x100

**Expected behavior (REQUIRED)**
 I expect it to stay 32x32 

**Defold version (REQUIRED):**
  - Version 1.4.3/1.4.4

### PR DESCRIPTION
**Pull request by matgis**
Fixes #7464

### User-facing changes
* Switching from auto to manual Size Mode on sprites or GUI objects will now set the manual size from the image size.
* Corrected the editor bounding box of sprites with manual Size Mode.

### CODE REVIEW

File: editor/src/clj/dev.clj [hunk 1]
```diff
@@ -164,6 +164,14 @@
...
   (-> (view-of-type console/ConsoleNode) (g/targets-of :lines) ffirst))
 
+(defn node-values [node-id & labels]
+  (g/with-auto-evaluation-context evaluation-context
+    (into {}
+          (map (fn [label]
+                 (let [value (g/node-value node-id label evaluation-context)]
+                   (pair label value))))
+          labels)))
+
```
Comment (matgis) [inline editor/src/clj/dev.clj:174]: Unrelated: Added a helper to `dev` for something I tend to do a lot.

File: editor/src/clj/dynamo/graph.clj [hunk 2]
```diff
@@ -160,8 +160,9 @@
          basis (is/basis system)
          id-generators (is/id-generators system)
          override-id-generator (is/override-id-generator system)
+         tx-data-context-map (or (:tx-data-context-map opts) {})
          metrics-collector (:metrics opts)
-         transaction-context (it/new-transaction-context basis id-generators override-id-generator metrics-collector)
+         transaction-context (it/new-transaction-context basis id-generators override-id-generator tx-data-context-map metrics-collector)
```
Comment (matgis) [inline editor/src/clj/dynamo/graph.clj:165]: The `:tx-data-context` field inside the `transaction-context` hosts an `atom` that can be used to communicate state changes inside the transaction. Previously we were using it to keep track of the node-ids of `ResourceNodes` created as a result of a property setter referring to a resource we haven't picked up yet to make sure we do not create duplicates. For example, if several sprites would refer to a missing material resource, we want all of them to refer to the same defective resource node, not create duplicate defective resource nodes.

This `:tx-data-context` atom is accessible from the `evaluation-context` supplied to all property `set` functions.

You can now supply an initial state for the map in the `:tx-data-context`, and we use it to host the set of property endpoints edited by the user using the Property Editor. This enables use to check inside a property `set` function if we're being invoked as a result of the user editing a particular property with the help of the new `properties/user-edit?` function.

File: editor/src/clj/editor/properties.clj [hunk 3]
```diff
@@ -460,15 +460,29 @@
...
+  individual [node-id prop-kw old-value new-value] vectors detailing the
+  low-level property set operations that will be performed as a result."
+  [property values]
+  (let [node-ids (:node-ids property)
+        prop-kws (:prop-kws property)
+        old-values (:values property)
+        new-values (if-some [from-fn (-> property :edit-type :from-type)]
+                     (mapv from-fn values)
+                     values)]
+    (mapv vector node-ids prop-kws old-values new-values)))
```
Comment (matgis) [inline editor/src/clj/editor/properties.clj:474]: Refactoring: A `property` map in the Property Editor can refer to one or more properties on various nodes. For example, when I select a `EmbeddedComponentNode` representing a Sprite component in the Property Editor, it is also presenting the properties from the embedded `SpriteNode`.

We need this sequence of resolved "set-operations" in `set-values` below, but also to calculate the `edited-endpoints`.

File: editor/src/clj/editor/properties.clj [hunk 4]
```diff
@@ -491,17 +505,35 @@
...
   ([property values]
    (set-values! property values (gensym)))
   ([property values op-seq]
    (when (not (read-only? property))
-     (let [evaluation-context (g/make-evaluation-context)]
+     (let [evaluation-context (g/make-evaluation-context)
+           set-operations (resolve-set-operations property values)
+           edited-endpoints (edited-endpoints set-operations)]
        (g/transact
+         {:tx-data-context-map {:edited-endpoints edited-endpoints}}
```
Comment (matgis) [inline editor/src/clj/editor/properties.clj:532]: I the future we might want to do this from the `form-view` as well, but currently only the Property Editor is used to edit the properties that call the `properties/user-edit?` function.

File: editor/src/clj/editor/sprite.clj [hunk 5]
```diff
@@ -375,12 +384,13 @@
...
-                                           (geom/make-aabb (Point3d. (- animation-width) (- animation-height) 0)
-                                                           (Point3d. animation-width animation-height 0)))
-                                         geom/empty-bounding-box)))
+  (output aabb AABB (g/fnk [size]
+                      (let [[^double width ^double height ^double depth] size
+                            half-width (* 0.5 width)
+                            half-height (* 0.5 height)
+                            half-depth (* 0.5 depth)]
+                        (geom/make-aabb (Point3d. (- half-width) (- half-height) (- half-depth))
+                                        (Point3d. half-width half-height half-depth)))))
```
Comment (matgis) [inline editor/src/clj/editor/sprite.clj:393]: Unrelated: I noticed the bounding boxes of sprites that use `:size-mode-manual` were not correct - the bounds were based on the texture size, causing box selection and framing to be wrong.

**Review by vlaaad — APPROVED** [review]
LGTM, but I have a question: is it possible to achieve the expected behavior by ensuring that during the loading we apply the size mode **before** applying the sizes? And in that case, what would be a better solution?

**Comment by matgis** [comment]
Yes, I do believe that would work. But we'd also have to ensure the texture is assigned after the `:size-mode` in order to avoid loading the image dimensions from the referenced `png` files, which is rather costly.

This could be avoided for sprites, but in the case of GUI nodes we create override nodes, and there would already be an existing texture hooked up to the original node.

I also feel being explicit like this avoids the potential future pitfall if something messes up the order the properties are applied.

**Pull request merged**